weblate/api/tests.py

```diff
@@ -203,6 +204,103 @@ def test_add_group(self):
             request={"group_id": group.id},
         )
 
+    def test_list_notifications(self):
+        response = self.do_request(
+            "api:user-notifications",
+            kwargs={"username": User.objects.filter(is_active=True).first().username},
+            method="get",
+            superuser=True,
+            code=200,
+        )
+        self.assertEqual(response.data["count"], 9)
+
+    def test_post_notifications(self):
+        self.do_request(
+            "api:user-notifications",
+            kwargs={"username": User.objects.filter(is_active=True).first().username},
+            method="post",
+            code=403,
+        )
+        self.do_request(
+            "api:user-notifications",
+            kwargs={"username": User.objects.filter(is_active=True).first().username},
+            method="post",
+            superuser=True,
+            code=201,
+            request={
+                "notification": "RepositoryNotification",
+                "scope": 10,
+                "frequency": 1,
+            },
+        )
+        self.assertEqual(Subscription.objects.count(), 10)
+
+    def test_get_notifications(self):
+        user = User.objects.filter(is_active=True).first()
+        self.do_request(
+            "api:user-notifications-details",
+            kwargs={"username": user.username, "subscription_id": -1},
+            method="get",
+            code=400,
+        )
+        self.do_request(
+            "api:user-notifications-details",
+            kwargs={
+                "username": user.username,
+                "subscription_id": Subscription.objects.filter(user=user).first().id,
+            },
+            method="get",
+            code=200,
+        )
+
+    def test_put_notifications(self):
+        user = User.objects.filter(is_active=True).first()
+        response = self.do_request(
+            "api:user-notifications-details",
+            kwargs={
+                "username": user.username,
+                "subscription_id": Subscription.objects.filter(user=user).first().id,
+            },
+            method="put",
+            superuser=True,
+            code=200,
+            request={
+                "notification": "RepositoryNotification",
+                "scope": 10,
+                "frequency": 1,
+            },
+        )
+        self.assertEqual(response.data["notification"], "RepositoryNotification")
+
+    def test_patch_notifications(self):
+        user = User.objects.filter(is_active=True).first()
+        response = self.do_request(
+            "api:user-notifications-details",
+            kwargs={
+                "username": user.username,
+                "subscription_id": Subscription.objects.filter(user=user).first().id,
```

Comment:
I'm affraid this is not reliable - there is no defined ordering on Subscription, so what is returned by `first()` depends on database ordering and thus is not something to rely on. It could easily happen that you already get subscription on "RepositoryNotification" and the patch doesn't work...

Reply:
Hmm. That's true. what would be a better test? filtering by user and notification != "RepositoryNotification" ?

Reply:
I'd probably just picked one of the notifications, for example "NewAnnouncementNotificaton" and use that for tests.

List of notifications subscribed by default for users is here:

https://github.com/WeblateOrg/weblate/blob/939f36c141244fc68fc73df5d6159e2792a7f129/weblate/accounts/data.py#L23-L33